core/src/avm2/verify.rs

```diff
@@ -0,0 +1,1126 @@
+                    let multiname = method
+                        .translation_unit()
+                        .pool_maybe_uninitialized_multiname(*index, &mut activation.context)?;
+
+                    if multiname.has_lazy_component() {
+                        return Err(Error::AvmError(verify_error(
+                            activation,
+                            "Error #1078: Illegal opcode/multiname combination.",
+                            1078,
+                        )?));
+                    }
+                }
+
+                AbcOp::AsType {
+                    type_name: name_index,
+                }
+                | AbcOp::Coerce { index: name_index } => {
+                    let multiname = method
+                        .translation_unit()
+                        .pool_maybe_uninitialized_multiname(*name_index, &mut activation.context)
+                        .unwrap();
+
+                    activation
+                        .domain()
+                        .get_class(&multiname, activation.context.gc_context)
+                        .unwrap()
+                        .ok_or_else(|| {
+                            Error::AvmError(
+                                verify_error(
+                                    activation,
+                                    &format!(
+                                        "Error #1014: Class {} could not be found.",
+                                        multiname.to_qualified_name(activation.context.gc_context)
+                                    ),
+                                    1014,
+                                )
+                                .expect("Error should construct"),
+                            )
+                        })?;
+                }
+
+                _ => {}
+            }
+
+            i += 1;
+        }
+
+        if !save_current_work {
+            worklist.pop();
+        }
+    }
+
+    Ok(())
+}
+
+fn optimize<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    method: &BytecodeMethod<'gc>,
+    code: &mut Vec<Op>,
+    jump_targets: HashSet<i32>,
+) {
+    // These make the code less readable
+    #![allow(clippy::manual_filter)]
+    #![allow(clippy::single_match)]
+
+    // This can probably be done better by recording the receiver in `Activation`,
+    // but this works since it's guaranteed to be set in `Activation::from_method`.
+    let this_value = activation.local_register(0);
+
+    let this_class = if let Some(this_class) = activation.subclass_object() {
+        if this_value.is_of_type(activation, this_class.inner_class_definition()) {
+            Some(this_class)
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    // Initial set of local types
+    let mut local_types = vec![None; method.body().unwrap().num_locals as usize];
+    local_types[0] = this_class;
+
+    // Invalidate local types if they should be invalidated
```

Comment:
This comment is a bit recursive.

It seems like this code currently just "invalidates" the remembered type of |this|, if that local is every modified. This approach doesn't really work for anything else.

Reply:
I plan to add parameter types here in a later PR (though that depends on #11954).

Reply:
Either way, how about documenting the purpose? Something like `A conservative logic to only allow type-based optimizations we're 100% sure about - only trust a parameter's type if its variable was never written to. We invalidate stored types if we encounter a register-setting opcode.`